integration/test_storage_http.py

```diff
@@ -0,0 +1,284 @@
+        response code is returned.
+        """
+        # Without secret, get a 400 error.
+        response = yield self.client._request(
+            "GET",
+            "http://127.0.0.1/upload_secret",
+        )
+        self.assertEqual(response.code, 400)
+
+        # With secret, we're good.
+        response = yield self.client._request(
+            "GET", "http://127.0.0.1/upload_secret", upload_secret=b"MAGIC"
+        )
+        self.assertEqual(response.code, 200)
+        self.assertEqual((yield response.content()), b"GOOD SECRET")
+
+
+class HttpTestFixture(Fixture):
+    """
+    Setup HTTP tests' infrastructure, the storage server and corresponding
+    client.
+    """
+
+    def _setUp(self):
+        self.tempdir = self.useFixture(TempDir())
+        self.storage_server = StorageServer(self.tempdir.path, b"\x00" * 20)
+        self.http_server = HTTPServer(self.storage_server, SWISSNUM_FOR_TEST)
+        self.client = StorageClient(
+            DecodedURL.from_text("http://127.0.0.1"),
+            SWISSNUM_FOR_TEST,
+            treq=StubTreq(self.http_server.get_resource()),
+        )
+
+
+class GenericHTTPAPITests(AsyncTestCase):
+    """
+    Tests of HTTP client talking to the HTTP server, for generic HTTP API
+    endpoints and concerns.
+    """
+
+    def setUp(self):
+        if PY2:
+            self.skipTest("Not going to bother supporting Python 2")
+        super(GenericHTTPAPITests, self).setUp()
+        self.http = self.useFixture(HttpTestFixture())
+
+    @inlineCallbacks
+    def test_bad_authentication(self):
+        """
+        If the wrong swissnum is used, an ``Unauthorized`` response code is
+        returned.
+        """
+        client = StorageClient(
+            DecodedURL.from_text("http://127.0.0.1"),
+            b"something wrong",
+            treq=StubTreq(self.http.http_server.get_resource()),
+        )
+        with self.assertRaises(ClientException) as e:
+            yield client.get_version()
+        self.assertEqual(e.exception.args[0], 401)
+
+    @inlineCallbacks
+    def test_version(self):
+        """
+        The client can return the version.
+
+        We ignore available disk space and max immutable share size, since that
+        might change across calls.
+        """
+        version = yield self.http.client.get_version()
+        version[b"http://allmydata.org/tahoe/protocols/storage/v1"].pop(
+            b"available-space"
+        )
+        version[b"http://allmydata.org/tahoe/protocols/storage/v1"].pop(
+            b"maximum-immutable-share-size"
+        )
+        expected_version = self.http.storage_server.get_version()
+        expected_version[b"http://allmydata.org/tahoe/protocols/storage/v1"].pop(
+            b"available-space"
+        )
+        expected_version[b"http://allmydata.org/tahoe/protocols/storage/v1"].pop(
+            b"maximum-immutable-share-size"
+        )
+        self.assertEqual(version, expected_version)
+
+
```

Comment:
Everything before this point in the module was moved from `src/allmydata/test/test_storage_http.py`, pretty much unchanged.

Reply:
Everything after this point is new code.